lib/spack/spack/build_environment.py

```diff
@@ -291,14 +291,29 @@ def set_build_environment_variables(pkg, env, dirty):
     # the given compiler
     compiler = pkg.compiler
     environment = compiler.environment
-    if 'set' in environment:
-        env_to_set = environment['set']
-        for key, value in iteritems(env_to_set):
-            env.set('SPACK_ENV_SET_%s' % key, value)
-            env.set('%s' % key, value)
-        # Let shell know which variables to set
-        env_variables = ":".join(env_to_set.keys())
-        env.set('SPACK_ENV_TO_SET', env_variables)
+
+    if issubclass(type(environment), dict):
+        # For backward compatibility
+        if 'set' in environment:
+            env_to_set = environment['set']
+            for key, value in iteritems(env_to_set):
+                env.set('SPACK_ENV_SET_%s' % key, value)
+                env.set('%s' % key, value)
+            # Let shell know which variables to set
+            env_variables = ":".join(env_to_set.keys())
+            env.set('SPACK_ENV_TO_SET', env_variables)
+    elif issubclass(type(environment), list):
+        for modification in environment:
+            env_cmd = modification[0]
+            env_cmd_args = modification[1:]
+            if env_cmd == 'set' and len(env_cmd_args) == 2:
+                env.set(*env_cmd_args)
```

Comment:
I disagree. A package knows better what it needs. @alalazo made a good point regarding this in https://github.com/spack/spack/issues/5019#issuecomment-342924735. But maybe you could give an example of when this can be helpful? Setting a variable in a wrapper makes a compiler work in an environment different from the environment, in which the configuration takes place, which is really confusing. See another story from @alalazo: https://github.com/spack/spack/issues/5019#issuecomment-340197220